Fix part of #10415: Extracts validators and tests related to subtopic model

Koushikk-24 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #10415.
2. This PR does the following: Modified tests and validators related files

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes made in this PR work correctly.
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

Koushikk-24 committed:
Create subtopic_validators.py

Extracted all the validators related to a given subtopic model type  subtopic_validators.py  from prod_validators.py. Making sure you name the validators file exactly as <folder-name-in-above-list>_validators.py

Koushikk-24 committed:
Created subtopic_validators_test.py

Extracted all the test related to the model type from prod_validators_test.py to subtopic_validators_test.py

Koushikk-24 committed:
Merge pull request #1 from Rupeshkoushik/Rupeshkoushik-patch-1

Create subtopic_validators.py

Koushikk-24 committed:
Merge pull request #2 from Rupeshkoushik/Rupeshkoushik-patch-2

Created subtopic_validators_test.py

Koushikk-24 committed:
Revert "Create subtopic_validators.py"

Koushikk-24 committed:
Merge pull request #4 from Rupeshkoushik/revert-1-Rupeshkoushik-patch-1

 "Create subtopic_validators.py"

Koushikk-24 committed:
Add files via upload

Koushikk-24 committed:
Update subtopic_validatiors_test.py

Koushikk-24 committed:
{{modified tests and validators}}

Koushikk-24 committed:
Update subtopic_validatiors_test.py

Koushikk-24 committed:
Update subtopic_validatiors_test.py

Koushikk-24 committed:
Merge remote-tracking branch 'upstream/develop' into develop

Koushikk-24 committed:
{{modified prod related files}}

Koushikk-24 committed:
Merge branch 'newPR' of https://github.com/Rupeshkoushik/oppia into newPR

Koushikk-24 commented:
@vojtechjelinek can you please explain what could be the reason for these errors I'm unable to get it

Koushikk-24 committed:
Update subtopic_validators.py

vojtechjelinek commented on core/domain/subtopic_validators.py:
```diff
@@ -0,0 +1,149 @@
+
+import python_utils
+
+(
+    base_models, exp_models, subtopic_models, user_models
+) = models.Registry.import_models([
+    models.NAMES.base_model, models.NAMES.subtopic, models.NAMES.user,models.NAMES.exploration
+])
```
Lines cannot be longer than 80 characters
```suggestion
(
    base_models, exp_models, 
    subtopic_models, user_models
) = models.Registry.import_models([
    models.NAMES.base_model, models.NAMES.subtopic, 
    models.NAMES.user, models.NAMES.exploration
])
```

Koushikk-24 replied:
Done!

vojtechjelinek commented on core/domain/prod_validators_test.py:
```diff
@@ -2345,6 +2346,283 @@ def test_invalid_entity_id_in_feedback_reference(self):
+                'of ContinuousComputationModel\', [u\'Entity id %s: '
+                'last stopped %s is greater than the current time\']]'
+            ) % (self.model_instance.id, self.model_instance.last_stopped_msec)]
+        self.run_job_and_check_output(
+            expected_output, sort=False, literal_eval=False)
+
+    def test_invalid_last_finished_msec(self):
+        current_time_msec = utils.get_current_time_in_millisecs()
+        self.model_instance.last_finished_msec = current_time_msec * 10.0
+        self.model_instance.update_timestamps()
+        self.model_instance.put()
+        expected_output = [
+            (
+                u'[u\'failed validation check for last finished check '
+                'of ContinuousComputationModel\', [u\'Entity id %s: '
+                'last finished %s is greater than the current time\']]'
+            ) % (
+                self.model_instance.id,
+                self.model_instance.last_finished_msec)]
+        self.run_job_and_check_output(
+            expected_output, sort=False, literal_eval=False)
+
+    def test_model_with_invalid_id(self):
+        model_with_invalid_id = job_models.ContinuousComputationModel(
+            id='invalid',
+            status_code=job_models.CONTINUOUS_COMPUTATION_STATUS_CODE_RUNNING,
+            last_started_msec=1, last_stopped_msec=10, last_finished_msec=20)
+        model_with_invalid_id.update_timestamps()
+        model_with_invalid_id.put()
+        expected_output = [
+            (
+                u'[u\'failed validation check for model id check of '
+                'ContinuousComputationModel\', '
+                '[u\'Entity id invalid: Entity id does not match '
+                'regex pattern\']]'
+            ), u'[u\'fully-validated ContinuousComputationModel\', 1]']
+        self.run_job_and_check_output(
+            expected_output, sort=True, literal_eval=False)
+
+
```
This shouldn't be added here, these are now moved to job_validators.py

Koushikk-24 replied:
Done!

vojtechjelinek reviewed (changes requested):
Thanks! Left two comments.

Koushikk-24 committed:
{{modified prod_validators_test.py}}

Koushikk-24 committed:
{{modified subtopic_validators.py}}

Koushikk-24 committed:
{{modified prod_validators_test.py and subtopic_validators.py}}

Koushikk-24 committed:
Merge branch 'develop' into newPR

Koushikk-24 committed:
{{modified prod_validators.py after conflicts}}

vojtechjelinek commented on core/domain/subtopic_validatiors_test.py:
```diff
@@ -0,0 +1,849 @@
+
+
+from constants import constants
+
+
```
Remove these empty lines

vojtechjelinek commented on core/domain/subtopic_validatiors_test.py:
```diff
@@ -0,0 +1,849 @@
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import datetime
+
+
```
Keep only one line here

vojtechjelinek commented on core/domain/prod_validators.py:
```diff
@@ -470,119 +469,342 @@ def _get_external_instance_custom_validation_functions(cls):
 
-class SubtopicPageModelValidator(base_model_validators.BaseModelValidator):
-    """Class for validating SubtopicPageModel."""
+class GeneralFeedbackThreadModelValidator(
+        base_model_validators.BaseModelValidator):
+    """Class for validating GeneralFeedbackThreadModels."""
```
This shouldn't be here, this is already placed in the feedback validators, the only changes in this file should be the removal of stuff that is now in subtopic validators, the same for the test file.

Koushikk-24 replied:
so the prod_validators.py file shouldn't contain any files related to subtopic model or the files in subtopic_validators.py

vojtechjelinek replied:
As of now the prod_validators should only contain exploration validators

vojtechjelinek replied:
The feedback validators were moved in #11340

vojtechjelinek replied:
There are still some in this file, that you need to remove

Koushikk-24 replied:
Done!

Koushikk-24 replied:
I have updated prod_validators.py file removing feedback validators

Koushikk-24 committed:
{{modified prod_validators and subtopic_validators_test.py}}

Koushikk-24 committed:
{{updated prod_validators.py }}

vojtechjelinek commented on core/domain/prod_validators.py:
```diff
@@ -212,7 +212,6 @@ def _get_external_id_relationships(cls, item):
     def _validate_first_published_msec(cls, item):
         """Validate that first published time of model is less than current
         time.
-
```
Remove this change, ditto below.

Koushikk-24 replied:
Should I add a blank line?

vojtechjelinek replied:
Yes, ditto in other places.

Koushikk-24 replied:
Done

vojtechjelinek commented on core/domain/prod_validators_test.py:
```diff
@@ -29,14 +29,13 @@
 from core.domain import story_domain
 from core.domain import story_services
-from core.domain import subtopic_page_domain
+
```
There should not be empty lines after you remove the imports, just remove the lines.

vojtechjelinek commented on core/domain/subtopic_validators.py:
```diff
@@ -0,0 +1,150 @@
+from core.domain import subtopic_page_domain
+from core.domain import subtopic_page_services
+from core.platform import models
+
```
Remove this empty line

Koushikk-24 committed:
{{updated}}

Koushikk-24 committed:
{{added blank lines}}

Koushikk-24 committed:
{{updated}}

Koushikk-24 commented:
@vojtechjelinek  the backend tests failed due to missing imports in subtopic_validators_test.py?

vojtechjelinek commented on core/domain/subtopic_validators.py:
```diff
@@ -0,0 +1,149 @@
+from core.platform import models
+import python_utils
+
+(
+    base_models, exp_models, subtopic_models, user_models
+) = models.Registry.import_models([
+    models.NAMES.base_model, models.NAMES.subtopic,
+    models.NAMES.user, models.NAMES.exploration
+])
```
Yo have a msitake here
```suggestion
(
    base_models, exp_models, 
    subtopic_models, user_models
) = models.Registry.import_models([
    models.NAMES.base_model, models.NAMES.exploration,
    models.NAMES.subtopic, models.NAMES.user, 
])
```

Rupeshkoushik replied:
Lines cannot be longer than 80 characters?

vojtechjelinek commented on core/domain/subtopic_validatiors_test.py:
```diff
@@ -0,0 +1,846 @@
+# coding: utf-8
```
This file has a wrong name: subtopic_validatiors_test.py → subtopic_validators_test.py

Koushikk-24 committed:
{{modified and renamed}}

Koushikk-24 committed:
{{modified subtopic_validators.py}}

Koushikk-24 committed:
{{resolved lint checks}}

vojtechjelinek commented on core/domain/subtopic_validators_test.py:
```diff
@@ -0,0 +1,846 @@
+        self.model_instance_1 = (
+            subtopic_models.SubtopicPageModel.get_by_id('1-1'))
+        self.model_instance_2 = (
+            subtopic_models.SubtopicPageModel.get_by_id('2-1'))
```
You missed copying the declaration of `self.job_class` here.

vojtechjelinek commented on core/domain/subtopic_validators_test.py:
```diff
@@ -0,0 +1,846 @@
+                '1-1-1'))
+        self.model_instance_2 = (
+            subtopic_models.SubtopicPageSnapshotMetadataModel.get_by_id(
+                '2-1-1'))
```
You missed copying the declaration of `self.job_class` here.

vojtechjelinek commented on core/domain/subtopic_validators_test.py:
```diff
@@ -0,0 +1,846 @@
+            subtopic_models.SubtopicPageSnapshotContentModel.get_by_id(
+                '2-1-1'))
+
+        self.job_class = (
+            prod_validation_jobs_one_off
+            .SubtopicPageSnapshotContentModelAuditOneOffJob)
```
Here it is okay, but you missed it in the other places.

Koushikk-24 replied:
Done!

vojtechjelinek commented on core/domain/subtopic_validators_test.py:
```diff
@@ -0,0 +1,846 @@
+                'subtopicpage-1-1-1'))
+        self.model_instance_2 = (
+            subtopic_models.SubtopicPageCommitLogEntryModel.get_by_id(
+                'subtopicpage-2-1-1'))
```
You missed copying the declaration of `self.job_class` here.

vojtechjelinek commented:
@Rupeshkoushik Left a few last comments, these should help with fixing the tests. Please be very careful when copying stuff since you shouldn't leave some parts out as you did here.

Koushikk-24 committed:
{{updated subtopic_validators_test.py}}

Koushikk-24 commented:
> @Rupeshkoushik Left a few last comments, these should help with fixing the tests. Please be very careful when copying stuff since you shouldn't leave some parts out as you did here.

Done!

vojtechjelinek commented on core/domain/subtopic_validators_test.py:
```diff
@@ -98,6 +98,10 @@ def setUp(self):
 
+        self.job_class = (
+            prod_validation_jobs_one_off
+            .SubtopicPageSnapshotContentModelAuditOneOffJob)
```
But these are not tests for `SubtopicPageSnapshotContentModel` but for `SubtopicPageModel`, ditto in other places.

Koushikk-24 replied:
I though they are same

Koushikk-24 committed:
Merge branch 'develop' into newPR

Koushikk-24 committed:
{{modified subtopic_validators_test.py}}

Koushikk-24 committed:
Merge branch 'newPR' of https://github.com/Rupeshkoushik/oppia into newPR

Koushikk-24 commented:
These errors are due to merge conflicts I will update it.

Koushikk-24 committed:
{{removed merge conflicts}}

vojtechjelinek commented on core/domain/subtopic_validators_test.py:
```diff
@@ -0,0 +1,858 @@
+            subtopic_models.SubtopicPageCommitLogEntryModel.get_by_id(
+                'subtopicpage-2-1-1'))
+
+        self.job_class = (
+            prod_validation_jobs_one_off
+            .SubtopicPageModelAuditOneOffJob)
```
Here you're working with other class

```suggestion
        self.job_class = (
            prod_validation_jobs_one_off
            .SubtopicPageCommitLogEntryModelAuditOneOffJob)
```

Koushikk-24 replied:
Done!
committed changes

vojtechjelinek commented on core/domain/subtopic_validators_test.py:
```diff
@@ -0,0 +1,858 @@
+            subtopic_models.SubtopicPageSnapshotContentModel.get_by_id(
+                '2-1-1'))
+
+        self.job_class = (
+            prod_validation_jobs_one_off
+            .SubtopicPageModelAuditOneOffJob)
```
Here you're working with other class

```suggestion
        self.job_class = (
            prod_validation_jobs_one_off
            .SubtopicPageSnapshotContentModelAuditOneOffJob)
```

vojtechjelinek commented on core/domain/subtopic_validators_test.py:
```diff
@@ -0,0 +1,858 @@
+            subtopic_models.SubtopicPageSnapshotMetadataModel.get_by_id(
+                '2-1-1'))
+
+        self.job_class = (
+            prod_validation_jobs_one_off
+            .SubtopicPageModelAuditOneOffJob)
```
Here you're working with other class

```suggestion
        self.job_class = (
            prod_validation_jobs_one_off
            .SubtopicPageSnapshotMetadataModelAuditOneOffJob)
```

vojtechjelinek commented on core/domain/subtopic_validators_test.py:
```diff
@@ -0,0 +1,858 @@
+        self.model_instance_2 = (
+            subtopic_models.SubtopicPageModel.get_by_id('2-1'))
+
+        self.job_class = (
+            prod_validation_jobs_one_off
+            .SubtopicPageModelAuditOneOffJob)
```
```suggestion
        self.job_class = (
            prod_validation_jobs_one_off.SubtopicPageModelAuditOneOffJob)
```

Koushikk-24 committed:
{{updated subtopic_validators_test.py}}

vojtechjelinek commented:
@Rupeshkoushik There are still backend tests failing, please make sure that you copied the code from one file to the other correctly and that you're on the current version of develop.

Koushikk-24 commented on core/domain/subtopic_validators_test.py:
```diff
@@ -0,0 +1,857 @@
+            '%s' % i,
+            'topic%s' % i,
+            'abbrev-%s' % chr(120 + i),
+            'description%s' % i) for i in python_utils.RANGE(3)]
```
It was this code which I left behind because it wasn't related to subtopic model 
The backend tests fail due to this?
rubrics = [
            skill_domain.Rubric(
                constants.SKILL_DIFFICULTIES[0], ['Explanation 1']),
            skill_domain.Rubric(
                constants.SKILL_DIFFICULTIES[1], ['Explanation 2']),
            skill_domain.Rubric(
                constants.SKILL_DIFFICULTIES[2], ['Explanation 3'])]
        skills = [
            skill_domain.Skill.create_default_skill(
                '%s' % i, 'skill%s' % i, rubrics)
            for i in python_utils.RANGE(9)]

        for skill in skills:
            skill_services.save_new_skill(self.owner_id, skill)

        stories = [story_domain.Story.create_default_story(
            '%s' % i,
            'title %d',
            'description %d' % i,
            '%s' % (python_utils.divide(i, 2)),
            'title-%s' % chr(97 + i)
        ) for i in python_utils.RANGE(6)]

        for story in stories:
            story_services.save_new_story(self.owner_id, story)

Koushikk-24 replied:
This was removed as in one of the tests if was mentioned undefined

Koushikk-24 replied:
@vojtechjelinek

vojtechjelinek replied:
What I meant was moving whole classes that relate to subtopic models, if there is some part in the class that doesn't relate to subtopics it needs to be kept.

The task basically is to move four validation classes and four test classes for them, nothing else.

sajalasati replied:
@Rupeshkoushik Is this issue resolved now? If yes then you can resolve it, else leave a comment about it explaining your concern (it's been quite a while now since last comment). Thanks!

Koushikk-24 replied:
Yes,the issue is resolved thanks!

Koushikk-24 commented on core/domain/subtopic_validators_test.py:
```diff
@@ -0,0 +1,857 @@
+        expected_output = [
+            u'[u\'fully-validated SubtopicPageSnapshotMetadataModel\', 4]']
+        self.run_job_and_check_output(
+            expected_output, sort=False, literal_eval=False)
```
def test_model_with_committer_id_migration_bot(self):
        self.model_instance_1.committer_id = feconf.MIGRATION_BOT_USER_ID
        self.model_instance_1.update_timestamps(update_last_updated_time=False)
        self.model_instance_1.put()
        expected_output = [
            u'[u\'fully-validated SubtopicPageSnapshotMetadataModel\', 3]'
        ]
        self.run_job_and_check_output(
            expected_output, sort=False, literal_eval=False)
    def test_model_with_pseudo_committer_id(self):
        self.model_instance_1.committer_id = self.PSEUDONYMOUS_ID
        self.model_instance_1.update_timestamps(update_last_updated_time=False)
        self.model_instance_1.put()
        expected_output = [
            u'[u\'fully-validated SubtopicPageSnapshotMetadataModel\', 3]'
        ]
        self.run_job_and_check_output(
            expected_output, sort=False, literal_eval=False)

sajalasati replied:
@Rupeshkoushik Does this still need any clarification from the reviewers? If yes, then please let us know in a follow up comment what is your particular concern here. Thanks!

Koushikk-24 replied:
No, it is resolved

Koushikk-24 committed:
{{added async keyword}}

Koushikk-24 committed:
{{added keyword async in question-backend-api.service.ts}}

Koushikk-24 committed:
{{updated subtopic_validators_test.py}}

Koushikk-24 committed:
Delete pip

Koushikk-24 committed:
{{removed unused imports in subtopic_validators_test}}

Koushikk-24 committed:
Merge remote-tracking branch 'upstream/develop' into newPR

Koushikk-24 committed:
Merge branch 'newPR' of https://github.com/Rupeshkoushik/oppia into newPR

Koushikk-24 committed:
{{updated lint checks in subtopic_validators_test.py}}

Koushikk-24 commented:
<img width="1437" alt="Screenshot 2021-01-08 at 7 58 03 PM" src="https://user-images.githubusercontent.com/73252546/104026850-89be1680-51ec-11eb-842a-de5ea9c79ad1.png">
errors im getting while performing tests locally @vojtechjelinek

vojtechjelinek commented:
@Rupeshkoushik If you compare the code that is currently on develop https://github.com/oppia/oppia/blob/develop/core/domain/prod_validators.py#L488-L610 and the code you have added to subtopic_validators.py, there is a difference (https://www.diffchecker.com/Ze1iFv7t). The same thing probably happened with the test file.

Koushikk-24 commented:
So my repo is behind the actual develop branch 
Means the subtopic_validators.py file should be changed to https://github.com/oppia/oppia/blob/develop/core/domain/prod_validators.py#L488-L610
And the subtopic_validators_test.py with reference to the current file of prod_validators_test.py file

vojtechjelinek commented:
@Rupeshkoushik Yes, basically, worth updating your repo to a current version.

Koushikk-24 committed:
{{updated subtopic_validators.py}}

Koushikk-24 commented:
@vojtechjelinek when https://github.com/oppia/oppia/blob/develop/core/domain/prod_validators_test.py#L1819-L2770 is compared to subtopic_validators_test.py there is no diff https://www.diffchecker.com/diff

vojtechjelinek commented on core/domain/subtopic_validators.py:
```diff
@@ -0,0 +1,157 @@
+from core.domain import subtopic_page_services
+from core.platform import models
+import python_utils
+(
+    base_models, exp_models,
+    subtopic_models, user_models
+) = models.Registry.import_models([
+    models.NAMES.base_model, models.NAMES.exploration,
+    models.NAMES.subtopic, models.NAMES.user
+])
```
You're missing `topic_models` here and put an empty line above it
```suggestion

(
    base_models, exp_models, subtopic_models, 
    topic_models, user_models
) = models.Registry.import_models([
    models.NAMES.base_model, models.NAMES.exploration, models.NAMES.subtopic,
    models.NAMES.topic, models.NAMES.user
])
```

vojtechjelinek commented on core/domain/subtopic_validators.py:
```diff
@@ -0,0 +1,157 @@
+    models.NAMES.base_model, models.NAMES.exploration,
+    models.NAMES.subtopic, models.NAMES.user
+])
+
```
Insert one more empty line here

Koushikk-24 replied:
Done!

vojtechjelinek reviewed (changes requested):
Look much better now, few notes.

Koushikk-24 committed:
{{updated lint checks for subtopic_validators.py}}

vojtechjelinek commented on core/domain/subtopic_validators.py:
```diff
@@ -0,0 +1,158 @@
+from core.domain import subtopic_page_services
+from core.platform import models
+import python_utils
+(
```
One more empty line should be above this (I already pointed this out in the previous comment, please read them more carefully).

Koushikk-24 replied:
Done!

Koushikk-24 committed:
{{added empty line}}

vojtechjelinek reviewed (approved):
LGTM!

vojtechjelinek commented:
@Rupeshkoushik Could you please merge from develop?

Koushikk-24 committed:
Merge branch 'newPR' into develop

Koushikk-24 commented:
> @Rupeshkoushik Could you please merge from develop?
Am I suppose to do in this way?

<img width="572" alt="Screenshot 2021-01-12 at 12 46 47 PM" src="https://user-images.githubusercontent.com/73252546/104282056-75bc3280-54d4-11eb-9c32-19dee95e6100.png">

vojtechjelinek commented:
@Rupeshkoushik You should be on `newPR` branch and do `git pull upstream develop` and then push.

Koushikk-24 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into newPR

Koushikk-24 commented:
> @Rupeshkoushik You should be on `newPR` branch and do `git pull upstream develop` and then push.

Done!

DubeySandeep commented on core/domain/prod_validators.py:
```diff
@@ -483,128 +481,3 @@ def _get_custom_validation_functions(cls):
-        return [
-            base_model_validators.ExternalModelFetcherDetails(
-                'subtopic_page_commit_log_entry_ids',
-                subtopic_models.SubtopicPageCommitLogEntryModel,
```
Remove the `subtopic_models` defined on top of this file (if it's not needed any more!).

Koushikk-24 replied:
done!

DubeySandeep replied:
Not done?

Koushikk-24 replied:
Hi, can you please check the file again?
All the changes required to this file were made before @DubeySandeep

DubeySandeep commented on core/domain/prod_validators_test.py:
```diff
@@ -50,11 +45,11 @@
 (
     collection_models, exp_models, feedback_models,
-    story_models, subtopic_models,
+    job_models, story_models, subtopic_models,
```
Any specific reason for this change?

Koushikk-24 replied:
This file should contain the remaining models ie exploration model
Should I remove all the models except exp_model?

DubeySandeep commented on core/domain/subtopic_validators.py:
```diff
@@ -0,0 +1,159 @@
+import python_utils
+
+(
+    base_models, exp_models, subtopic_models,
```
Do we need `exp_models` here?

Koushikk-24 replied:
No its not required

DubeySandeep commented on core/domain/subtopic_validators.py:
```diff
@@ -0,0 +1,159 @@
+
+(
+    base_models, exp_models, subtopic_models,
+    topic_models, user_models
```
Ditto for `topic_models`

Koushikk-24 replied:
This causes error as line 69 gets undefined

DubeySandeep commented on core/domain/subtopic_validators.py:
```diff
@@ -0,0 +1,159 @@
+# coding: utf-8
```
Can you please assign this file to @aks681 in the `New structure` project inside CODOWNER files? 
`core/domain/subtopic_*.py @aks681` [line 398]

/cc @sajalasati

Koushikk-24 replied:
Yes assigned

DubeySandeep replied:
(Sorry I had mentioned incorrect number line before*)

DubeySandeep commented on core/domain/subtopic_validators_test.py:
```diff
@@ -0,0 +1,1005 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Unit tests for core.domain.prod_validators."""
```
Update the docstring*

Koushikk-24 replied:
Updated to subtopic_validators_test

DubeySandeep commented on core/domain/subtopic_validators_test.py:
```diff
@@ -0,0 +1,1005 @@
+USER_NAME = 'username'
+
+(
+    collection_models, exp_models, feedback_models,
+    story_models, subtopic_models,
+    suggestion_models, topic_models, user_models
```
Remove unwanted models form here

Koushikk-24 replied:
Done!

DubeySandeep reviewed (changes requested):
@Rupeshkoushik, Thanks for the PR, I took a quick pass and left a few comments PTAL!

Koushikk-24 committed:
Merge remote-tracking branch 'upstream/develop' into develop

Koushikk-24 committed:
Merge branch 'develop' into newPR

Koushikk-24 committed:
Merge branch 'async-update' of https://github.com/Rupeshkoushik/oppia into develop

Koushikk-24 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

Koushikk-24 committed:
removed unwanted models

Koushikk-24 committed:
removed unwanted models

Koushikk-24 committed:
removed unwanted models

Koushikk-24 committed:
removed unwanted models

DubeySandeep commented:
@Rupeshkoushik, Can you please reply to all the comments above if it's resolved? ([following the wiki guidelines](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#address-review-comments-until-all-reviewers-give-lgtm-looks-good-to-me))

Koushikk-24 committed:
Merge branch 'newPR' of https://github.com/Rupeshkoushik/oppia into newPR

Koushikk-24 committed:
Update subtopic_validators.py

DubeySandeep commented:
@Rupeshkoushik, Backend tests are failing on this PR, PTAL!

Rupeshkoushik commented:
@DubeySandeep PTAL!

Koushikk-24 committed:
updated`

Koushikk-24 committed:
backup

Koushikk-24 commented:
<img width="489" alt="Screenshot 2021-01-27 at 11 45 28 AM" src="https://user-images.githubusercontent.com/73252546/105965081-03894780-60a9-11eb-9d75-839d6a024b9d.png">
@DubeySandeep all tests were passed when I run it in locally

sajalasati commented:
> @DubeySandeep all tests were passed when I run it in locally

Hi @Rupeshkoushik! The passed checks message is **only** for lint checks run by the `pre_commit_linter` on the files in which you made changes. It does not check anything else. You still have failing backend and linter checks to be fixed.
- The Linter check on github runs over all the files. You can look at the error by clicking on details. You can fix the particular error by including the `./core/domain/subtopic_validators*.py @vojtechjelinek` in a new line below [this](https://github.com/oppia/oppia/blob/6a5283fbd8f59cf74909531a63138a6278efc445/.github/CODEOWNERS#L446) line in the CODEOWNERS file.
- Similarly check which backend tests are failing and fix them too. Please take a look at our wiki on [how to run backend tests locally](https://github.com/oppia/oppia/wiki/Running-Tests#server-side-tests), so you can the run the backend tests you have written locally before pushing. 
I hope this is helpful for you!

Koushikk-24 commented:
@sajalasati ./core/domain/subtopic_validators_test.py is also not listed in CODEOWNER file can I add it?
and the backend test was failing with subtopic_validators_test file when I run the backend test locally It say all checks were passed
EDIT: Now I'm getting backend errors when the tests were run locally
<img width="538" alt="Screenshot 2021-01-27 at 4 18 39 PM" src="https://user-images.githubusercontent.com/73252546/105982784-c7141680-60bd-11eb-9035-531e0d40deba.png">

sajalasati commented:
> @sajalasati ./core/domain/subtopic_validators_test.py is also not listed in CODEOWNER file can I add it?

Sorry I didn't see this before - @DubeySandeep has made a [comment](https://github.com/oppia/oppia/pull/11582/files#r556011850) on this earlier so I request you to follow that instead.
In general it is important to address all the previous comments (by replying to them with a "Done." atleast) before you ask for any re-review. I can see a lot of unaddressed comments currently, so PTAL!

> and the backend test was failing with subtopic_validators_test file when I run the backend test locally It say all checks were passed

Okay, then it might be some other check that is getting affected by your changes. Can you check which test is failing on github by clicking on details and let us know if case you are unable to solve it? (It might not be directly useful, but can you also make sure your branch is up-to-date with upstream/develop?) Thanks!

Koushikk-24 commented:
> @Rupeshkoushik, Backend tests are failing on this PR, PTAL!

Done!

Koushikk-24 committed:
Merge remote-tracking branch 'upstream/develop' into newPR

Koushikk-24 committed:
merged and resolved conflicts

Koushikk-24 committed:
added subtopic_validators.py

Koushikk-24 committed:
updated

Koushikk-24 committed:
updated codeowner file

Koushikk-24 committed:
updated codeowner files

Koushikk-24 commented:
> > @sajalasati ./core/domain/subtopic_validators_test.py is also not listed in CODEOWNER file can I add it?
> 
> Sorry I didn't see this before - @DubeySandeep has made a [comment](https://github.com/oppia/oppia/pull/11582/files#r556011850) on this earlier so I request you to follow that instead.
> In general it is important to address all the previous comments (by replying to them with a "Done." atleast) before you ask for any re-review. I can see a lot of unaddressed comments currently, so PTAL!
> 
> > and the backend test was failing with subtopic_validators_test file when I run the backend test locally It say all checks were passed
> 
> Okay, then it might be some other check that is getting affected by your changes. Can you check which test is failing on github by clicking on details and let us know if case you are unable to solve it? (It might not be directly useful, but can you also make sure your branch is up-to-date with upstream/develop?) Thanks!

Updated branch with upstream/develop and added subtopic file in CODEOWNER file

Koushikk-24 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

Koushikk-24 committed:
Merge remote-tracking branch 'upstream/develop' into develop

Koushikk-24 committed:
updated prod_validators

Koushikk-24 committed:
Updated subtopic_valdiators

sajalasati commented:
@Rupeshkoushik Can you look once again into the failing backend tests? I guess you haven't written (or have removed) the tests covering lines mentioned in the [log](https://github.com/oppia/oppia/pull/11582/checks?check_run_id=1791212635#step:8:6676), hence getting a code coverage error. PTAL!

sajalasati reviewed (commented):
@Rupeshkoushik Left a few comments related to some of your older comments too. PTAL! 
Once you reply to them and fix the failing code coverage in backend tests, please feel free to re-assign the PR to a reviewer (so that they know it's their turn to take a look). Thanks!

Koushikk-24 commented:
> @Rupeshkoushik Can you look once again into the failing backend tests? I guess you haven't written (or have removed) the tests covering lines mentioned in the [log](https://github.com/oppia/oppia/pull/11582/checks?check_run_id=1791212635#step:8:6676), hence getting a code coverage error. PTAL!

Done! I have updated

Koushikk-24 committed:
backend tests resolved

sajalasati commented:
@Koushikk-24 Don't forget to assign the next person who has to review your PR and un-assign yourself. I will leave the task of reviewing to codeowners only here. Thanks!

vojtechjelinek reviewed (approved):
LGTM!

Koushikk-24 committed:
removed unwanted models

Koushikk-24 committed:
Merge remote-tracking branch 'upstream/develop' into develop

Koushikk-24 committed:
Merge remote-tracking branch 'upstream/develop' into develop

Koushikk-24 commented:
> @Koushikk-24 Don't forget to assign the next person who has to review your PR and un-assign yourself. I will leave the task of reviewing to codeowners only here. Thanks!

Done!

DubeySandeep commented on .github/CODEOWNERS:
```diff
@@ -396,6 +396,7 @@
 # Story project.
 /core/controllers/story*.py @aks681
 /core/domain/story*.py @aks681
+/core/domain/subtopic_*.py @aks68
```
Invalid userID, please add correct the userID @aks681 (Missing 1 at the end.)

Koushikk-24 replied:
Updated!

DubeySandeep commented on .github/CODEOWNERS:
```diff
@@ -396,6 +396,7 @@
 # Story project.
 /core/controllers/story*.py @aks681
 /core/domain/story*.py @aks681
+/core/domain/subtopic_*.py @aks68
```
Change it to `/core/domain/subtopic*.py` and move this to the `Topic project` section.

Koushikk-24 replied:
Done!

DubeySandeep commented on core/domain/prod_validators.py:
```diff
@@ -28,8 +28,6 @@
-from core.domain import subtopic_page_services
 from core.platform import models
 import feconf
 import python_utils
```
Remove `subtopic_models` from line 39*

Koushikk-24 replied:
Hi, can you please check the file again?
All the changes required to this file were made before @DubeySandeep

DubeySandeep replied:
I checked the [file](https://github.com/oppia/oppia/blob/8df803bb9a096fc72ef6e84564294f04d88e7574/core/domain/prod_validators.py) again and I think `subtopic_models` defined on line 39 is no more needed, can you please check and remove (if required)?

Koushikk-24 replied:
Done!

DubeySandeep commented on core/domain/subtopic_validators.py:
```diff
@@ -0,0 +1,159 @@
+# coding: utf-8
+#
+# Copyright 2020 The Oppia Authors. All Rights Reserved.
```
2021*

Koushikk-24 replied:
Done

DubeySandeep commented on core/domain/subtopic_validators_test.py:
```diff
@@ -0,0 +1,1003 @@
+# coding: utf-8
+#
+# Copyright 2019 The Oppia Authors. All Rights Reserved.
```
2021*

Koushikk-24 replied:
Done!

DubeySandeep commented on core/domain/subtopic_validators_test.py:
```diff
@@ -0,0 +1,1003 @@
+USER_NAME = 'username'
+
+(
+    story_models, subtopic_models,
```
Do we need `story_models` here?

Koushikk-24 replied:
Yes, without this can lead to a failure in backend tests. @DubeySandeep

DubeySandeep replied:
I don't see any place where `story_models` is used in this file. can you please share the error log?

Koushikk-24 replied:
<img width="1180" alt="Screenshot 2021-02-05 at 8 49 27 AM" src="https://user-images.githubusercontent.com/73252546/106986766-10d1c080-6792-11eb-990d-683634069b65.png">
The same applies for removal of topic_model in subtopic_validators.py which was mentioned above @DubeySandeep

DubeySandeep replied:
The errors are related to story_domain class not defined (I believe you have removed the story_domain import), but I'm asking to remove story_models. Can you please check after only removing the story_models?

Koushikk-24 replied:
Yes, you are right the change was due to removal of story_domain and also story_models was also removed 
@DubeySandeep PTAL

DubeySandeep reviewed (changes requested):
@Koushikk-24, I took a pass and left a few comments PTAL!

Koushikk-24 committed:
updated docstring and CODEOWNER file

U8NWXD commented:
@DubeySandeep PTAL. Your review was re-requested

DubeySandeep commented on core/domain/prod_validators_test.py:
```diff
@@ -451,10 +446,11 @@ def test_maximum_of_ten_errors_are_emitted(self):
         exp_models.ExplorationModel.get_by_id('0').delete(
             self.user_id, '', [])
 
-        self.process_and_flush_pending_tasks()
```
Why this change?

Koushikk-24 replied:
The change was made after updating local branch(this was updated in master branch)

DubeySandeep replied:
Is this an expected change? That's what I wanted to check, if not then please undo these changes! :) [Ditto below]

Koushikk-24 replied:
Yes, the changes are expected

DubeySandeep commented on core/domain/prod_validators_test.py:
```diff
@@ -1453,6 +1449,7 @@ def setUp(self):
                 'property_name': 'title',
                 'new_value': 'New title'
             })], 'Changes.')
+        self.process_and_flush_pending_tasks()
```
Ditto?

Koushikk-24 replied:
The change was made after updating local branch(this was updated in master branch)

DubeySandeep commented:
@Koushikk-24, I've left a few comments above and [here](https://github.com/oppia/oppia/pull/11582#discussion_r570459379), PTAL!

Koushikk-24 committed:
removed unwanted models

U8NWXD commented:
@Koushikk-24 if you're ready for DubeySandeep to re-review, please leave a comment like `@DubeySandeep PTAL!`. Oppiabot will then automatically assign them

Koushikk-24 commented:
@DubeySandeep PTAL!

DubeySandeep commented:
@Koushikk-24, I've left follow-up comments in the above comment thread, PTAL!

Koushikk-24 commented:
@DubeySandeep I have followed all the comments and also resolved ,PTAL

U8NWXD commented:
@Koushikk-24 can you merge from develop? You need to merge in the updated backend test config file, otherwise github won't let you merge

sajalasati commented:
@Koushikk-24 Can you please merge from develop once?

Koushikk-24 committed:
Merge branch 'develop' into newPR

Koushikk-24 commented:
> @Koushikk-24 Can you please merge from develop once?

Yes, I have merged from develop just now

Koushikk-24 committed:
Merge remote-tracking branch 'upstream/develop' into develop

Koushikk-24 committed:
Merge remote-tracking branch 'upstream/develop' into newPR

hardikkat24 commented:
@Koushikk-24  I think you have not pushed the merged code. Please push the merged code.

hardikkat24 commented:
@Koushikk-24 The backend tests are failing because when you create new tests, you should add them to the list of shards. Read more [here](https://groups.google.com/g/oppia-dev/c/k6ppgwvq3hI). I would suggest you to add the test to shard 8.

Koushikk-24 commented:
> @Koushikk-24 The backend tests are failing because when you create new tests, you should add them to the list of shards. Read more [here](https://groups.google.com/g/oppia-dev/c/k6ppgwvq3hI). I would suggest you to add the test to shard 8.

Thanks! I will update it

Koushikk-24 committed:
updated file name in shards

Koushikk-24 committed:
Delete question-backend-api.service.ts

Koushikk-24 committed:
updated file name in shards

Koushikk-24 committed:
Merge remote-tracking branch 'upstream/develop' into develop

Koushikk-24 committed:
Merge remote-tracking branch 'origin' into newPR

Koushikk-24 committed:
updated shards.json

U8NWXD commented:
@Koushikk-24 can you take a look at the typescript failures? Here's the link: https://app.circleci.com/pipelines/github/oppia/oppia/18873/workflows/7fbf9354-4754-4705-be13-1db39c710b5c/jobs/106699.

kevintab95 reviewed (dismissed):
Thanks @Koushikk-24! Why is `core/templates/domain/question/question-backend-api.service.ts` deleted?

Koushikk-24 commented:
> Thanks @Koushikk-24! Why is `core/templates/domain/question/question-backend-api.service.ts` delete

The file dosen't belong to this PR as this PR contains only the files related to the subtopic.

sajalasati commented:
> > Thanks @Koushikk-24! Why is `core/templates/domain/question/question-backend-api.service.ts` delete
> 
> The file dosen't belong to this PR as this PR contains only the files related to the subtopic.

@Koushikk-24 You should not be deleting/modifying files lying outside the scope of this PR. I think you should merge from develop once, and that should solve the problem. Let us know if you face any issues in doing so. Thanks!

Koushikk-24 committed:
removing merge conflicts  Conflicts:

U8NWXD commented:
Reopened at @Koushikk-24's request

Koushikk-24 committed:
updated lint checks

Koushikk-24 committed:
resolved lint checks

Koushikk-24 committed:
resolved backend tests`

Koushikk-24 committed:
solved backend failures`

Koushikk-24 committed:
resolved backend tests

Koushikk-24 commented:
@kevintab95 PTAL

U8NWXD commented:
@Koushikk-24 it looks like you need to add some backend tests for your code. The [coverage report](https://github.com/oppia/oppia/pull/11582/checks?check_run_id=2186167141) lists the lines that aren't covered:

```
Name                                                                             Stmts   Miss  Cover   Missing
--------------------------------------------------------------------------------------------------------------
...
core/domain/playthrough_issue_registry.py                                           31      0   100%
core/domain/prod_validation_jobs_one_off.py                                        358      2    99%   113-114
core/domain/prod_validators.py                                                     126     66    48%   75, 79-88, 92-95, 124, 139, 143, 163, 175-181, 221-226, 234, 245, 249, 269, 285-288, 292-301, 305-319, 327, 331, 363-367, 383-388, 414-435, 447-465, 477, 483
core/domain/question_domain.py                                                     422      0   100%
core/domain/question_fetchers.py                                                    48      0   100%
...
--------------------------------------------------------------------------------------------------------------
TOTAL                                                                            38817     68    99%
Coverage failure: total of 100 is less than fail-under=100
```

Please don't just write tests for those lines though. Think about all the ways the code you wrote could be used, and write tests for those cases

DubeySandeep commented:
Closing this PR because of lack of activity! @Koushikk-24 feel free to re-open this PR once it's ready for merge! :)
